src/go/configgenerator/filterconfig/filter_generator.go

```diff
@@ -165,6 +169,60 @@ func MakeFilterGenerators(serviceInfo *ci.ServiceInfo) ([]*FilterGenerator, erro
 	return filterGenerators, nil
 }
 
+func updateProtoDescriptor(serviceInfo *ci.ServiceInfo, descriptorBytes []byte) []byte {
+	// To support specifying custom http rules in service config.
+	// api-compiler (or CLI: gcloud endpoints service deploy) has following behaviours:
+	// * If a "google.api.http" annotation is specified in a method in the proto,
+	//   service config yaml did not specify one, api-compiler will copy it out
+	//   to the normalized service config.
+	// * If a htpp.rule is specified in the service config, it will override
+	//   the one from the annotation, used it in the normalized service config.
+	// The normalized service config is used in this function. Its http rules will be
+	// copied into the method.options in the proto descriptor to be used by the
+	// envoy grpc_jsontranscoder.
+	ruleMap := make(map[string]*ahttp.HttpRule)
+	for _, rule := range serviceInfo.ServiceConfig().GetHttp().GetRules() {
+		ruleMap[rule.GetSelector()] = rule
+	}
+	apiMap := make(map[string]bool)
+	for _, apiName := range serviceInfo.ApiNames {
+		apiMap[apiName] = true
+	}
+
+	fds := &descpb.FileDescriptorSet{}
+	if err := proto.Unmarshal(descriptorBytes, fds); err != nil {
+		glog.Error("failed to unmarshal protodescriptor, error: %v", err)
+		return descriptorBytes
+	}
+
+	for _, file := range fds.GetFile() {
+		for _, service := range file.GetService() {
+			apiName := fmt.Sprintf("%s.%s", file.GetPackage(), service.GetName())
+
+			// Only modify the API in the serviceInfo.ApiNames.
+			if _, ok := apiMap[apiName]; !ok {
+				continue
+			}
+
+			for _, method := range service.GetMethod() {
+				sel := fmt.Sprintf("%s.%s", apiName, method.GetName())
+				if rule, ok := ruleMap[sel]; ok {
+					json, _ := util.ProtoToJson(rule)
+					glog.Info("Set http.rule: ", json)
+					proto.SetExtension(method.GetOptions(), ahttp.E_Http, rule)
+				}
+			}
+		}
+	}
+
+	newData, err := proto.Marshal(fds)
+	if err != nil {
+		glog.Error("failed to marshal protodescriptor, error: %v", err)
+		return descriptorBytes
```

Comment:
Return the error. This looks a critical failure.

Reply:
Done